Fixes #10850 - handled errors when missing template when rendering widget

iNecas opened the pull request:
How to reproduce:
- install a plugin with a widget (such as foreman discovery),
- configure the widget from the plugin to be displayed
- uninstall the plugin

iNecas committed:
Fixes #10850 - handled errors when missing template when rendering widget

ohadlevy commented on app/views/dashboard/index.html.erb:
```diff
@@ -11,15 +11,20 @@
+
+                <%= begin
+                      render_widget(widget)
+                    rescue ActionView::MissingTemplate => e; e.class
```
should widgit should be extended to to contain their source? this way we can join the current plugins and widgets instead of rescuing? /cc @tbrisker

iNecas replied:
Possibly, but I liked this, because it was
1. less work
2. letting the user to fix their dashboard: if they already have customized dashboard, I though it would be friendlier for the user just reporting the issue to them.

On the other hand, that doesn't solve issues, when the dashboard renames the template: in that case we should not save the paths to the database, but rather load that from the widget defitntion, witch would identify the problem.

unorthodoxgeek replied:
I think that the begin/rescue clause should be in the render_widget helper method. this has no place in a view.
also, something seems funky with the indentation...

iNecas replied:
good point

iNecas replied:
updated

unorthodoxgeek commented:
now looking good :+1:

dLobatog commented on app/helpers/dashboard_helper.rb:
```diff
@@ -15,6 +15,8 @@ def dashboard_actions
   def render_widget(widget)
     render(:partial => widget.template, :locals => widget.data)
+  rescue ActionView::MissingTemplate => e
+    _("Missing template for widget '%s'. The widget or a plugin providing it was probably removed") % widget.name
```
I think "Widget cannot be found" is a bit more precise and less hesitant, easier to translate (vs "probably removed").

 Can you raise a Foreman::Exception so that there's an ERF code for that, so that when this is seen, people can search for it and find possible causes for it in the wiki?

iNecas replied:
I don't want to reraise: the point of this bug was that, if you uninstall the plugin, you start getting 500 in dashboard, and there was not easy way how to fix that. With this fix, you get mostly empty widget space, with the message why it's that. Reraising the error would take us to the start

iNecas replied:
@eLobato I need your reply before proceeding any further

dLobatog commented:
@iNecas Tested, it does rescue well with discovery widgets at least. A test that raises ActionView::MissingTemplate in render_widget and ensures we get an ERF would be great.

Thank you!

unorthodoxgeek commented:
@iNecas - please fix that rubucop warning (how I hate those), I'm good with this code, as soon as @eLobato  approves this should be merge-able :+1:

iNecas commented:
I've updated the code (the rubocop should be happy now) + added more data to the error message: the template path that was not found might be also useful

dLobatog commented:
[test]

dLobatog commented on app/helpers/dashboard_helper.rb:
```diff
@@ -15,6 +15,8 @@ def dashboard_actions
   def render_widget(widget)
     render(:partial => widget.template, :locals => widget.data)
+  rescue ActionView::MissingTemplate
+    _("Missing template '%{template}' for widget '%{widget}'. The widget or a plugin providing it was probably removed") % { :widget => widget.name, :template => widget.template }
```
Can you raise a `Foreman::Exception` that is shown in an alert here instead? I think we can do a better job explaining possible causes by filling the wiki for a given ERF rather than just mention a cause (without the solution) in the message itself.

ares replied:
@inecas' in outdated comment above:

> I don't want to reraise: the point of this bug was that, if you uninstall the plugin, you start getting 500 in dashboard, and there was not easy way how to fix that. With this fix, you get mostly empty widget space, with the message why it's that. Reraising the error would take us to the start

I agree that we shouldn't raise any exception here.

domcleal replied:
Seems reasonable.

Nitpick: add a full stop to the end of the second sentence, since there's one on the first.

iNecas replied:
I've removed the sentence itself, as in theory there might be more issues that caused problem, and we can leave the explanation to the errors wiki

dLobatog commented:
Sorry, raising the exception defeats the whole purpose of this bug. But if you substitute the line for this

```
::Foreman::Exception.new(N_("Missing template '%{template}' for widget '%{widget}'.") % { :widget => widget.name, :template => widget.template })
```

we don't get a 500, instead the message plus an error code are printed. I just think it could be useful to have the error code in our wiki and explain how to solve the issue there.

![](http://i.imgur.com/M8JvmBC.png)

dLobatog commented:
@iNecas ping?

iNecas commented:
@eLobato rebased and updated based on comments (nice trick with the exception!), I've modified slightly your suggestion (the message params needed to be passed as separate argument to always calculate the same error code for the exception)

domcleal commented on app/helpers/dashboard_helper.rb:
```diff
@@ -15,6 +15,8 @@ def dashboard_actions
   def render_widget(widget)
     render(:partial => widget.template, :locals => widget.data)
+  rescue ActionView::MissingTemplate
+    ::Foreman::Exception.new(N_("Missing template '%{template}' for widget '%{widget}'."), :widget => widget.name, :template => widget.template)
```
Widget names should be extracted for i18n, so I think you can safely do `_(widget.name)` here to potentially get a translated version.

iNecas replied:
Good idea! Updated

dLobatog commented:
Unrelated Katello failure, [test]

domcleal commented:
Merged as 7fdaccd, thanks @iNecas.
